### ISSUE TITLE
SQLDroid on JDK (non-Android) platforms.

### ISSUE THREAD

**Issue by ergouser** [SQLDroid/SQLDroid#73]
SQLDroid can be used on standard JDK platforms through the use of the JNA/Compatibility layer.  This needs to be documented and brought into the build/test system.

Basic procedure:

Build a jar, say "SQLDroidCompat.jar" from:
compat/android
add that, the jna.jar and the appropriate .so, .dll, etc.for you platform from the SQLite website to your project (as well as the SQLDroid jar).

Works on Windows (32 & 64) and Linux.

**Comment by jhannes** [comment]
Another approach could be to introduce an interface between the SQLDroidXXX classes and the android.database-classes. This has already been begun with the SQLiteDatabase class. This way, the android.database classes and the compat-classes could be two different implementations of the same interface and no manual build steps would be needed.

This doesn't solve the problem of having to have the appropriate .so or .dll. Xerial solves this with https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/core/NativeDB.java, but at a cost of several megabytes of extra weight to the jar-file. But perhaps an alternative build could use this approach.

**Comment by ergouser** [comment]
@jhannes I'm not sure that I understand what you have in mind.  SQLite.java is just a wrapper for the dll/so via JNA.  Last time I looked, Xerial was doing the same thing with JNI but they bundle all the native code in the jar so it doesn't look as though there's a native piece.  The SQLite "database" is just a file on the disk, the "engine" is the native C code, everything else is wrappers.  Truly native Java would require that the entire engine be re-implemented in Java.

Much of the code in android.database is just a (fairly thin) layer on top of the SQLite class.  There's not much there and I suspect that the Android code - which I haven't looked at - is not much different.  Since the Android code is available, that could probably be used with minimal changes (or patch up missing features, etc.)

This approach also means that SQLDroid can continue to progress in the Android space (where it seems to be most used) without consideration for any other platform.

**Comment by jhannes** [comment]
@ergouser It's hard to explain design. I've made an experimental branch to demonstrate what I mean:

https://github.com/SQLDroid/SQLDroid/tree/jdk-experimental-poc

I have not added tests for this yet, but in this commit, the only thing missing is the library. If you use this commit outside android, or with the JDBC url jdbc:jnasqlite:<filename>, it will use the implementation previously under compat (moved to src/main/java/org/sqldroid/sqlite in this branch).

This is just a quick PoC. What do you think?

**Comment by ergouser** [comment]
I'm not convinced of the benefit.  It adds complexity to the Android SQLDroid by introducing an new layer.  The code in the jna layer is basically the same as ever but in a new package.  Is there something that makes you uncomfortable about having the code in android.database and to prefer the jna/android structure?  (I spend a lot of time emulating existing packages, which is how we run code originally written for AWT/Swing on Android and GWT so I may well have a skewed view of the world).

**Comment by jhannes** [comment]
AFAIK it would create huge problems to have android.* packages in the jar file when deploying on Android, so if the jar should contain emulating code, it needs to have a different package name.

I agree that it adds a layer of complexity, but in the case of your usage, it just makes an existing layer of complexity visible. :-) The advantage is that this way all the code can be under one build management system.

As an example, I discovered that a few of the Cursor methods had changed signatures subtly, so the contrib-code didn't compile.

**Comment by ergouser** [comment]
Agreed on not deploying android.packages to Android.  We keep the compatibility jar with the JDK code and add the SQLDroid jar as needed.  If there's a goal of putting the native code into the jar, then one jar is probably not a good plan. I don't have any problems with either approach.  From the JDK side I like your approach better, my concern is whether this is an acceptable direction for the Android side.

**Comment by ergouser** [comment]
Unrelated, I wish I could remember who to credit for the SQLite.java file - that's not originally mine.

**Comment by jhannes** [comment]
I think the change makes the code somewhat more logically structured for Android as well. This way, the "top" of the code - the JDBC implementation - is clearly separate from the bottom - adapting to the android API.

@donv What do you think of this PoC restructure?

**Comment by donv** [comment]
I don't have the full overview of the change, but what I have seen of the PoC makes sense.

I'd like to build and test both variants in travis so they don't decay.

I want the Android package to be as lean as possible:  no extra classes for non-Android use in the Android distribution jar.

**Comment by jhannes** [comment]
@donv: It's a pretty large diff online because of the file moves. The essence of the change is [SQLDroidConnection.createDatabase](https://github.com/SQLDroid/SQLDroid/blob/a6fe19475fc7c363c3470e6045e219b071292f21/src/main/java/org/sqldroid/SQLDroidConnection.java#L147) In addition, [SQLiteDatabase](https://github.com/SQLDroid/SQLDroid/blob/a6fe19475fc7c363c3470e6045e219b071292f21/src/main/java/org/sqldroid/SQLiteDatabase.java) has been made into an interface and the implementation renamed AndroidSQLiteDatabase.

@ergouser The easiest way to fulfill the need to avoid non-Android classes is probably to subclass SQLDroidDriver. This means that you'd need to use another class for Driver.registerDriver, e.g. org.sqldroid.jna.JnaSQLDroidDriver. Do you think that's acceptable?

**Comment by ergouser** [comment]
@jhannes Having a different Driver class would be a problem for us because we use exactly the same code on JDK and Android.  It's not an impossible problem - with jarjar(https://github.com/shevek/jarjar) many things are possible.  Keeping the android.database classes means that the JDK version has another jar, but I'm not sure that's a problem and could be an advantage if we decide to package the native code.  

How badly do you really dislike the android.database compatibility classes?

**Comment by jhannes** [comment]
@ergouser Okay, then it has to be the same driver class. Reflection can be used to support missing classes.

I really dislike that some of the code is not part of the build. I also see the potential for the JDK classes providing better test support. I've been unable to collect coverage from the current test suite, but the JDK-classes could help with that.

Also, it seems like currently, the JDK version isn't distributed. That means that probably only you are using it, but others could, if we could build and release it at the same time as the android version. One way of doing this is to use the maven concept of classifiers.

**Comment by ergouser** [comment]
@jhannes I agree on testing. I'm not sure why two jars vs. one jar makes a difference for distribution or anything else.  The jna.jar and a native library are also required.
